TLM/TLM/UI/SubTools/LaneConnectorTool.cs

```diff
@@ -15,626 +15,651 @@
+            for (int i = 0; i < 8; i++) {
+                ushort segmentId = node.GetSegment(i);
+                if (segmentId == 0)
+                    continue;
+                bool isEndNode = NetManager.instance.m_segments.m_buffer[segmentId].m_endNode == nodeId;
+                Vector3 offset = NetManager.instance.m_segments.m_buffer[segmentId].FindDirection(segmentId, nodeId) * offsetMultiplier;
+                NetInfo.Lane[] lanes = NetManager.instance.m_segments.m_buffer[segmentId].Info.m_lanes;
+                uint laneId = NetManager.instance.m_segments.m_buffer[segmentId].m_lanes;
+                NetInfo info = NetManager.instance.m_segments.m_buffer[segmentId].Info;
+                List<NetInfo.Lane> sortedLanes = lanes.Where(l => (l.m_laneType & LaneConnectionManager.LANE_TYPES) != NetInfo.LaneType.None
+                    && (l.m_vehicleType & LaneConnectionManager.VEHICLE_TYPES) != VehicleInfo.VehicleType.None).OrderBy(l => l.m_position).ToList();
+                int[] laneGroups = new int[sortedLanes.Count()];
+                if ((node.m_flags & NetNode.Flags.Junction) == NetNode.Flags.None) {
+                    for (byte laneIndex = 1; laneIndex < sortedLanes.Count && laneId != 0; laneIndex++) {
+                        laneGroups[laneIndex] = laneGroups[laneIndex - 1];
+                        NetInfo.Lane laneInfo = sortedLanes[laneIndex];
+                        NetInfo.Lane prevLaneInfo = sortedLanes[laneIndex - 1];
+                        if ((laneInfo.m_vehicleType & VehicleInfo.VehicleType.Car) == VehicleInfo.VehicleType.None) {
+                            laneGroups[laneIndex]++;
+                        }
+                        if (laneInfo.m_finalDirection != prevLaneInfo.m_finalDirection) {
+                            laneGroups[laneIndex]++;
+                        } else {
+                            var gap = laneInfo.m_position - prevLaneInfo.m_position - (0.5f * prevLaneInfo.m_width) - (0.5f * laneInfo.m_width);
+                            if (gap > 1 ) {
+                                laneGroups[laneIndex]++;
+                            }
+                        }
+                    }
+                }
+                for (byte laneIndex = 0; laneIndex < lanes.Length && laneId != 0; laneIndex++) {
+                    NetInfo.Lane laneInfo = lanes[laneIndex];
+                    if ((laneInfo.m_laneType & LaneConnectionManager.LANE_TYPES) != NetInfo.LaneType.None &&
+                        (laneInfo.m_vehicleType & LaneConnectionManager.VEHICLE_TYPES) != VehicleInfo.VehicleType.None) {
+
+                        Vector3? pos = null;
+                        bool isSource = false;
+                        bool isTarget = false;
+                        if (connManager.GetLaneEndPoint(segmentId, !isEndNode, laneIndex, laneId, laneInfo, out isSource, out isTarget, out pos)) {
+                            int sortedLaneIndex = sortedLanes.IndexOf(laneInfo);
```

Comment:
how about something like `byte LaneConnectionManager.GetLaneEndGroup(ushort segmentId, ref NetSegment segment, ushort nodeId, ref NetNode node, byte laneIndex, NetInfo.Lane laneInfo, uint laneId)`?

Reply:
(we may not need all of the parameters I mentioned ^^)